pkg/blocks/ttps.go

```diff
@@ -119,228 +118,180 @@ func reduceIndentation(b []byte, n int) []byte {
-		Description string            `yaml:"description"`
-		Environment map[string]string `yaml:"env,flow,omitempty"`
-		Steps       []yaml.Node       `yaml:"steps,omitempty,flow"`
-		ArgSpecs    []args.Spec       `yaml:"args,omitempty,flow"`
-	}
-
-	var tmpl TTPTmpl
-	if err := node.Decode(&tmpl); err != nil {
-		return err
-	}
-
-	t.Name = tmpl.Name
-	t.Description = tmpl.Description
-	t.Environment = tmpl.Environment
-	t.ArgSpecs = tmpl.ArgSpecs
+// error: An error if any part of the validation fails, otherwise nil.
+func (t *TTP) Validate(execCtx TTPExecutionContext) error {
+	logging.L().Info("[*] Validating Steps")
 
-	// Check for and handle a mitre node
-	var mitreNode *yaml.Node
-	for i := 0; i < len(node.Content)-1; i += 2 {
-		keyNode := node.Content[i]
-		if keyNode.Value == "mitre" {
-			mitreNode = node.Content[i+1]
-			break
-		}
+	// validate MITRE mapping
+	if t.MitreAttackMapping != nil && len(t.MitreAttackMapping.Tactics) == 0 {
+		return fmt.Errorf("TTP '%s' has a MitreAttackMapping but no Tactic is defined", t.Name)
 	}
 
-	if mitreNode != nil {
-		if err := mitreNode.Decode(&t.MitreAttackMapping); err != nil {
+	for _, step := range t.Steps {
+		stepCopy := step
+		if err := stepCopy.Validate(execCtx); err != nil {
 			return err
 		}
-		// if we have a MitreAttackMapping, ensure there's a tactic
-		if len(t.MitreAttackMapping.Tactics) == 0 {
-			return fmt.Errorf("TTP '%s' has a MitreAttackMapping but no Tactic is defined", t.Name)
-		}
-	}
-
-	return t.decodeSteps(tmpl.Steps)
-}
-
-func (t *TTP) decodeSteps(steps []yaml.Node) error {
-	for stepIdx, stepNode := range steps {
-		decoded := false
-		// these candidate steps are pointers, so this line
-		// MUST be inside the outer step loop or horrible things will happen
-		// #justpointerthings
-		stepTypes := []Step{NewBasicStep(), NewFileStep(), NewSubTTPStep(), NewEditStep(), NewFetchURIStep(), NewCreateFileStep()}
-		for _, stepType := range stepTypes {
-			err := stepNode.Decode(stepType)
-			if err == nil && !stepType.IsNil() {
-				// Must catch bad steps with ambiguous types, such as:
-				// - name: hello
-				//   file: bar
-				//   ttp: foo
-				//
-				// we can't use KnownFields to solve this without a massive
-				// refactor due to https://github.com/go-yaml/yaml/issues/460
-				if decoded {
-					return fmt.Errorf("step #%v has ambiguous type", stepIdx+1)
-				}
-				logging.L().Debugw("decoded step", "step", stepType)
-				t.Steps = append(t.Steps, stepType)
-				decoded = true
-			}
-		}
-
-		if !decoded {
-			return fmt.Errorf("Step #%v does not match any supported step type", stepIdx+1)
-		}
 	}
-
+	logging.L().Info("[+] Finished validating steps")
 	return nil
 }
 
-func (t *TTP) setWorkingDirectory() error {
-	if t.WorkDir != "" {
-		return nil
+func (t *TTP) chdir() (func(), error) {
+	// note: t.WorkDir may not be set in tests but should
+	// be set when actualy using `ttpforge run`
+	if t.WorkDir == "" {
```

Comment:
slight refactor of directory handling - we now pass it in execution context instead of as part of the step itself - various aspects have been updated accordingly